ard_controller/ard_controller.ino

```diff
@@ -15,45 +25,274 @@ struct JointAngleStruct {
+        if (temporary_packet_data.checksum == 0xFF - (calculated_checksum % 256)) {
+          // good packet
+          calculated_checksum = 0x00;
+          preamble_counter = PREAMBLE_LENGTH;
+
+          temporary_packet_data.checksum_error = false;
+          temporary_packet_data.packet_available = true;
+
+          validated_packet_data.left_hip = temporary_packet_data.left_hip;
+          validated_packet_data.left_knee = temporary_packet_data.left_knee;
+          validated_packet_data.right_hip = temporary_packet_data.right_hip;
+          validated_packet_data.right_knee = temporary_packet_data.right_knee;
+          validated_packet_data.left_shoulder = temporary_packet_data.left_shoulder;
+          validated_packet_data.left_elbow = temporary_packet_data.left_elbow;
+          validated_packet_data.right_shoulder = temporary_packet_data.right_shoulder;
+          validated_packet_data.right_elbow = temporary_packet_data.right_elbow;
+          validated_packet_data.checksum = temporary_packet_data.checksum;
+          validated_packet_data.checksum_error = temporary_packet_data.checksum_error;
+          validated_packet_data.packet_available = temporary_packet_data.packet_available;
+          validated_packet_data.data_request = temporary_packet_data.data_request;
+
+          sr_state = INIT;
+        }
+        else {
+          temporary_packet_data.checksum_error = true;
+          temporary_packet_data.packet_available = false;
+          sr_state = INIT;
+        }
+        break;
+      default:
+        sr_state = INIT;
+        break;
+      }
   }
 }
 
+void sensor_data_response() {
+  int raw_sum = 0;
+  int checksum;
+  long angle, goal;
+  int b1, b2;
+
+  // send preamble
+  for (int i = 0; i < PREAMBLE_LENGTH; i++) {
+    Serial1.write((byte)255);
+  }
+
+  // send imu and joint angle data
+  for (int i = 0; i < NUM_JOINTS + NUM_IMU_AXES; i++) {
+    angle = random(360);
+    raw_sum += angle;
+    b1 = angle / 256 > 0 ? 255 : angle;
+    b2 = angle / 256 > 0 ? angle % 255 : 0;
+
+    Serial1.write((byte)b1);
+    Serial1.write((byte)b2);
+  }
+
+  // send at_goal flag
+  goal = random(2);
+  raw_sum += goal;
+  Serial1.write((byte)goal);
+
+  // send checksum
+  checksum = 255 - raw_sum % 256;
+  Serial1.write((byte)checksum);
+}
+
 void setup() {
-  // put your setup code here, to run once:
   Serial.begin(115200);
   Serial1.begin(115200);
-
-  Timer1.attachInterrupt(serial_ISR).setFrequency(1).start();
 }
 
 void loop() {
-  int data;
-  if (Serial1.available() > 0) {
-    data = Serial1.read();
-    Serial.println("pong");
-  }
+  rx_processor();
+  if (validated_packet_data.packet_available) {
+    dump_validated_packet_data();
+    validated_packet_data.packet_available = false;
 
-  if (write_flag) {
-    Serial.println("ping");
-    write_flag = false;
+    if (validated_packet_data.data_request) {
+      sensor_data_response();
+    }
+    else {
+      // we received a new joint angle goal; get to it
```

Comment:
I am assuming this is where we will actuate the joints?

Reply:
yep